mesh/mesh.go

```diff
@@ -281,87 +234,88 @@ func (msh *Mesh) processValidityUpdates(ctx context.Context, logger log.Log, upd
-				if !block.Data {
-					continue
-				}
-				if err := blocks.UpdateValid(dbtx, block.Header.ID, block.Valid); err != nil {
-					return err
-				}
-			}
-		}
+	if changed == 0 {
 		return nil
-	}); err != nil {
-		return fmt.Errorf("save block validity: %w", err)
 	}
-	msh.validityUpdates = nil
+	revert := changed.Sub(1)
+	msh.logger.With().Info("reverting state",
+		log.Context(ctx),
+		log.Uint32("revert_to", revert.Uint32()),
+	)
+	if err := msh.executor.Revert(ctx, revert); err != nil {
+		return fmt.Errorf("revert state to layer %v: %w", revert, err)
+	}
+	if err := layers.UnsetAppliedFrom(msh.cdb, revert.Add(1)); err != nil {
+		return fmt.Errorf("unset applied layer %v: %w", revert.Add(1), err)
+	}
+	msh.setLatestLayerInState(revert)
 	return nil
 }
 
-// ProcessLayer performs fairly heavy lifting: it triggers tortoise to process the full contents of the layer (i.e.,
-// all of its blocks), then to attempt to validate all unvalidated layers up to this layer. It also applies state for
-// newly-validated layers.
-func (msh *Mesh) ProcessLayer(ctx context.Context, layerID types.LayerID) error {
-	logger := msh.logger.WithContext(ctx).WithFields(log.Stringer("processing", layerID))
-
-	// pass the layer to tortoise for processing
-	msh.trtl.TallyVotes(ctx, layerID)
-
+// ProcessLayer reads latest consensus results and ensures that vm state
+// is consistent with results.
+// It is safe to call after optimistically executing the block.
+func (msh *Mesh) ProcessLayer(ctx context.Context, lid types.LayerID) error {
 	msh.mu.Lock()
 	defer msh.mu.Unlock()
 
-	logger.Debug("processing layer")
+	msh.logger.With().Debug("processing layer",
+		log.Context(ctx),
+		log.Uint32("layer_id", lid.Uint32()),
+	)
 
-	// set processed layer even if later code will fail, as that failure is not related
-	// to the layer that is being processed
-	if err := msh.setProcessedLayer(logger, layerID); err != nil {
-		return err
-	}
+	msh.trtl.TallyVotes(ctx, lid)
 
-	updated := msh.trtl.Updates()
-	if err := msh.processValidityUpdates(ctx, logger, updated); err != nil {
+	if err := msh.setProcessedLayer(lid); err != nil {
 		return err
 	}
-
-	// mesh can't skip layer that failed to complete
-	from := msh.LatestLayerInState().Add(1)
-	to := layerID
-	if from == msh.MissingLayer() {
-		to = msh.maxProcessedLayer
+	results := msh.trtl.Updates()
+	pending := msh.pendingUpdates.min != 0
+	if len(results) > 0 {
+		msh.pendingUpdates.min = types.MinLayer(msh.pendingUpdates.min, results[0].Layer)
+		msh.pendingUpdates.max = types.MaxLayer(msh.pendingUpdates.max, results[len(results)-1].Layer)
 	}
-
-	if !to.Before(from) {
-		if err := msh.pushLayersToState(ctx, logger, from, to); err != nil {
+	if next := msh.LatestLayerInState() + 1; next < msh.pendingUpdates.min {
+		msh.pendingUpdates.min = next
+		pending = true
+	}
+	if pending {
+		var err error
+		results, err = msh.trtl.Results(msh.pendingUpdates.min, msh.pendingUpdates.max)
```

Comment:
i am confused by Updates() and Results() api.
the code is counting on `msh.pendingUpdates.min` to remember the min layer from the last tortoise update.
and if the mesh somehow was't able to apply the block from call, code asks for trtl for more prior updates that mesh didn't persist. these two apis are very similar.

would it make sense to create have a channel btwn from trtl->mesh that pushes the min updated layer each time?
that way mesh can query results btwn min(instate+1, trtl min updated layer) and the latest verified layer. then we'll just need Results()

Reply:
> latest verified layer

max is not verified layer, as we want to get also latest optimistic results with computed opinion

maybe i can make updates to return *Range instead of []result.Layer

Reply:
i will look into that in https://github.com/spacemeshos/go-spacemesh/issues/4425